update.go

```diff
@@ -512,3 +524,259 @@ func rollbackUpdatedFiles(currentApplicationDir string, updateFiles []updateFile
 	}
 	return
 }
+
+const ZipArchiveExtension = ".zip"
+
+var TarGzArchiveExtensions = []string{".tgz", ".tar.gz"}
+
+type versionFilesLoader struct {
+	version                Version
+	updateConfig           *UpdateConfig
+	getReplacementFileInfo func(loadedFilename string) (ReplacementFile, error)
+	destDir                string
+	createOrdinaryFiles    bool // create temp files otherwise USE CAREFULLY files could replace each other in ordinary naming
+}
+
+func (vfl versionFilesLoader) loadVersionFiles() ([]updateFile, error) {
+	assetsFilenames := vfl.version.getAssetsFilenames()
+	archivesFilenames := make([]string, 0)
+	for i, filename := range assetsFilenames {
+		isArchive := false
+		for _, ext := range append(TarGzArchiveExtensions, ZipArchiveExtension) {
+			ok, err := filepath.Match("*"+ext, filename)
+			if err != nil {
+				return nil, err
+			}
+			if ok {
+				isArchive = true
+				break
+			}
+		}
+		if isArchive {
+			archivesFilenames = append(archivesFilenames, filename)
+			continue
+		}
+		assetsFilenames[i-len(archivesFilenames)] = filename
+	}
+	assetsFilenames = assetsFilenames[:len(assetsFilenames)-len(archivesFilenames)]
+
+	updateFilesInfo, err := vfl.loadUpdateFilesFromSource(assetsFilenames)
+	if err != nil {
+		return nil, err
+	}
+	archivesUpdateFilesInfo, err := vfl.loadUpdateFilesFromSourceArchives(archivesFilenames)
+	if err != nil {
+		return nil, err
+	}
+	return append(updateFilesInfo, archivesUpdateFilesInfo...), nil
+}
+
+func (vfl versionFilesLoader) loadUpdateFilesFromSourceArchives(archivesFilenames []string) ([]updateFile, error) {
+	updateFilesInfo := make([]updateFile, 0)
+	for _, archive := range archivesFilenames {
+		fExt := filepath.Ext(archive)
+		fPath, err := vfl.loadUpdateFileFromSource(archive)
+		if err != nil {
+			return nil, err
+		}
+		var ufi []updateFile
+		if fExt == ZipArchiveExtension {
+			ufi, err = vfl.unpackZipArchive(fPath)
+
+		} else {
+			ufi, err = vfl.unpackTarGzArchive(fPath)
+		}
+		if err != nil {
+			return nil, err
+		}
+		updateFilesInfo = append(updateFilesInfo, ufi...)
+	}
+	return updateFilesInfo, nil
+}
+
+func (vfl versionFilesLoader) unpackZipArchive(archiveFPath string) ([]updateFile, error) {
+	updateFilesInfo := make([]updateFile, 0)
+	zR, err := zip.OpenReader(archiveFPath)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		zRCloseErr := zR.Close()
+		if err != nil && zRCloseErr != nil {
+			err = fmt.Errorf("%v; zip close error: %v", err, zRCloseErr)
+		}
+		if err == nil {
+			err = zRCloseErr
+		}
+	}()
+
+	for _, file := range zR.File {
+		if file.FileInfo().IsDir() {
+			continue
+		}
+		fName := filepath.Base(file.Name)
+		replacementFileInfo, err1 := vfl.getReplacementFileInfo(fName)
+		if err1 != nil || replacementFileInfo.PreventFileLoading {
```

Comment:
Why return if `|| replacementFileInfo.PreventFileLoading`? I thought it's for one particular file from archive, not all archive

Reply:
My mistake, in my first realization this code was in function